host/frontend/host-orchestrator/instancemanager.go

```diff
@@ -14,22 +14,171 @@
 
 func (m *InstanceManager) CreateCVD(req *CreateCVDRequest) (*Operation, error) {
 	if err := validateRequest(req); err != nil {
 		return nil, err
 	}
+	// This logic isn't complete yet, it's work in progress.
+	go func() {
+		err := m.fetchCVDHandler.Download(req.FetchCVDBuildID)
+		if err != nil {
+			log.Printf("error downloading fetch_cvd: %v\n", err)
+		}
+	}()
 	return &Operation{}, nil
 }
 
 func validateRequest(r *CreateCVDRequest) error {
 	if r.BuildInfo == nil ||
 		r.BuildInfo.BuildID == "" ||
 		r.BuildInfo.Target == "" ||
-		r.FetchCVDBuildID == "" ||
-		r.BuildAPIAccessToken == "" {
+		r.FetchCVDBuildID == "" {
 		return NewBadRequestError("invalid CreateCVDRequest", nil)
 	}
 	return nil
 }
+
+type FetchCVDHandler struct {
+	dir                string
+	fetchCVDDownloader FetchCVDDownloader
+	mutex              sync.Mutex
+}
+
+func NewFetchCVDHandler(dir string, fetchCVDDownloader FetchCVDDownloader) *FetchCVDHandler {
+	return &FetchCVDHandler{dir, fetchCVDDownloader, sync.Mutex{}}
+}
+
+func (h *FetchCVDHandler) Download(buildID string) error {
+	h.mutex.Lock()
+	defer h.mutex.Unlock()
+	exist, err := h.exist(buildID)
+	if err != nil {
+		return err
+	}
+	if exist {
+		return nil
+	}
+	fileName := BuildFetchCVDFileName(h.dir, buildID)
+	f, err := os.Create(fileName)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	if err := h.fetchCVDDownloader.Download(f, buildID); err != nil {
+		if removeErr := os.Remove(fileName); err != nil {
+			return fmt.Errorf("%w; %v", err, removeErr)
+		}
+		return err
+	}
+	return nil
+}
+
+func (h *FetchCVDHandler) exist(buildID string) (bool, error) {
+	if _, err := os.Stat(BuildFetchCVDFileName(h.dir, buildID)); err == nil {
+		return true, nil
+	} else if errors.Is(err, os.ErrNotExist) {
+		return false, nil
+	} else {
+		return false, err
+	}
+}
+
+func BuildFetchCVDFileName(dir, buildID string) string {
+	return fmt.Sprintf("%s/fetch_cvd_%s", dir, buildID)
+}
+
+type FetchCVDDownloader interface {
+	Download(dst io.Writer, buildID string) error
+}
+
+// Downloads fetch_cvd binary from Android Build.
+type ABFetchCVDDownloader struct {
+	client *http.Client
+	url    string
+}
+
+func NewABFetchCVDDownloader(client *http.Client, URL string) *ABFetchCVDDownloader {
+	return &ABFetchCVDDownloader{client, URL}
+}
+
+func (d *ABFetchCVDDownloader) Download(dst io.Writer, buildID string) error {
+	signedURL, err := d.getSignedURL(buildID)
+	if err != nil {
+		return err
+	}
+	req, err := http.NewRequest("GET", signedURL, nil)
+	if err != nil {
+		return err
+	}
+	res, err := d.client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+	if res.StatusCode != http.StatusOK {
+		return d.parseErrorResponse(res.Body)
+	}
+	_, err = io.Copy(dst, res.Body)
+	return err
+}
+
+func (d *ABFetchCVDDownloader) getSignedURL(buildID string) (string, error) {
+	url := fmt.Sprintf("%s/android/internal/build/v3/builds/%s/%s/attempts/%s/artifacts/%s/url?redirect=false",
+		d.url,
+		buildID,
```

Comment:
For safety this should probably be url-escaped.

Reply:
done